File: frontend/src/crypto/transparency.js
/**
 * Recompute a peer's key history here, in the browser.
 *
 * The server returns a `chain_ok` of its own, and it is worth being explicit about why
 * that field is ignored: a relay reporting on whether it has tampered with its own log is
 * not evidence of anything. The hashes are the evidence, and they only mean something once
 * this side has recomputed them.
 *
 * The chain gives one property, precisely. Each entry commits to the entry before it, so
 * the sequence cannot be shortened, reordered or edited after the fact without every
 * following hash failing. A relay can still stay silent -- decline to show an entry, or
 * serve an older head -- and no log fixes that. What it can no longer do is substitute a
 * key and produce a history in which that key was always there.
 *
 * And it says nothing about first contact. With no earlier state to compare against there
 * is nothing to detect, which is why the safety-number comparison in verification.js
 * remains the thing that establishes who you are talking to. This makes a *change*
 * visible; it does not make an introduction trustworthy.
 *
 * Kept byte-compatible with `backend/app/transparency.py::entry_hash` — the domain
 * separator, the field order and the 4-byte length prefixes all have to match, or every
 * chain fails to verify here and the warning means nothing.
 */
import { base64ToBytes, concatBytes, encoder } from "./bytes.js";

const DOMAIN = encoder.encode("prahari-key-transparency-v1");
const ZERO32 = new Uint8Array(32);

const hex = (bytes) =>
  [...bytes].map((byte) => byte.toString(16).padStart(2, "0")).join("");

/** Big-endian, matching Python's `int.to_bytes(width, "big")`. */
function intBytes(value, width) {
  const out = new Uint8Array(width);
  let remaining = BigInt(value);
  for (let index = width - 1; index >= 0; index -= 1) {
    out[index] = Number(remaining & 0xffn);
    remaining >>= 8n;
  }
  return out;
}

/**
 * Lengths precede the variable-length fields. Without them, moving bytes from the end of
 * one key to the start of the next would leave the concatenation — and so the hash —
 * unchanged, letting two different bundles share a digest.
 */
async function entryHash(entry, previousHash) {
  const parts = [DOMAIN, previousHash ?? ZERO32];
  const fields = [
    encoder.encode(entry.user_id),
    base64ToBytes(entry.ed25519_public_key),
    base64ToBytes(entry.x25519_public_key),
    base64ToBytes(entry.ml_kem_encapsulation_key),
  ];
  for (const field of fields) {
    parts.push(intBytes(field.length, 4), field);
  }
  parts.push(intBytes(entry.seq, 8));
  return new Uint8Array(await crypto.subtle.digest("SHA-256", concatBytes(...parts)));
}

/**
 * @returns {Promise<{ok: boolean, reason: string|null, head: string|null, changes: number}>}
 */
export async function verifyKeyHistory(history) {
  const entries = history?.entries ?? [];
  if (!entries.length) {
    return { ok: false, reason: "This account has never published a key.", head: null, changes: 0 };
  }

  let previousHash = null;
  for (const [index, entry] of entries.entries()) {
    if (entry.seq !== index + 1) {
      return {
        ok: false,
        reason: `The history skips from ${index} to ${entry.seq} — an entry is missing.`,
        head: null,
        changes: 0,
      };
    }
    const expectedPrev = previousHash ? hex(previousHash) : null;
    if ((entry.prev_hash ?? null) !== expectedPrev) {
      return {
        ok: false,
        reason: `Entry ${entry.seq} does not follow the one before it.`,
        head: null,
        changes: 0,
      };
    }
    const recomputed = await entryHash({ ...entry, user_id: history.user_id }, previousHash);
    if (hex(recomputed) !== entry.entry_hash) {
      return {
        ok: false,
        reason: `Entry ${entry.seq} has been altered since it was published.`,
        head: null,
        changes: 0,
      };
    }
    previousHash = recomputed;
  }

  return {
    ok: true,
    reason: null,
    head: hex(previousHash),
    // The first publish is an account existing, not a key changing.
    changes: entries.length - 1,
  };
}

/**
 * Compare a verified head against the one this device recorded last time.
 *
 * A head that moved is not an alarm by itself — people legitimately rotate keys and set up
 * new devices. It is an alarm when nobody told you, which is why this reports the
 * transition rather than a verdict.
 */
export function describeHeadChange(previousHead, currentHead) {
  if (!previousHead) return { changed: false, first: true };
  return { changed: previousHead !== currentHead, first: false };
}

File: backend/app/transparency.py
"""The key transparency log.

What this closes, precisely. A key bundle is signed by its owner's Ed25519 identity key,
so the relay has never been able to *fabricate* one -- it cannot produce the signature.
What it could do until now was quieter: serve an older bundle than the current one, or
serve a bundle belonging to a different identity, and nothing in the record would
disagree. The account row held one bundle and no history, so "this is their key" and
"this became their key an hour ago" looked identical.

Appending every publish to a per-user hash chain makes the second class of answer
checkable. Each entry commits to the one before it, so removing, reordering or editing an
entry invalidates every hash after it, and a client that has seen an earlier chain state
can prove the relay has changed its story.

Say the limit as plainly as the property. A log makes a *changed* answer detectable; it
does nothing about a *first* answer. On first contact there is no prior state to compare
against, so the safety-number comparison in crypto/verification.js remains the only thing
that establishes who you are actually talking to. Transparency turns silent substitution
into visible history -- it does not remove the need to check a fingerprint once.
"""
from __future__ import annotations

import hashlib

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from .models import KeyBundleRecord

#: Domain separator. Without it, a digest computed here could be replayed as a Merkle leaf
#: from blockchain/anchor.py, which hashes with a `\x00` prefix over arbitrary bytes.
_DOMAIN = b"prahari-key-transparency-v1"


def entry_hash(
    *,
    prev_hash: bytes | None,
    user_id: str,
    seq: int,
    ed25519_public_key: bytes,
    x25519_public_key: bytes,
    ml_kem_encapsulation_key: bytes,
) -> bytes:
    """Commit to one bundle and to everything published before it.

    Lengths are written before the variable-length fields rather than relying on
    concatenation. Without them, two different bundles could produce identical input --
    the classic boundary ambiguity where moving bytes from the end of one field to the
    start of the next leaves the joined string unchanged.
    """
    digest = hashlib.sha256()
    digest.update(_DOMAIN)
    digest.update(prev_hash or b"\x00" * 32)
    for field in (user_id.encode("utf-8"), ed25519_public_key, x25519_public_key, ml_kem_encapsulation_key):
        digest.update(len(field).to_bytes(4, "big"))
        digest.update(field)
    digest.update(seq.to_bytes(8, "big"))
    return digest.digest()


async def latest_record(db: AsyncSession, user_id: str) -> KeyBundleRecord | None:
    return (
        await db.execute(
            select(KeyBundleRecord)
            .where(KeyBundleRecord.user_id == user_id)
            .order_by(KeyBundleRecord.seq.desc())
            .limit(1)
        )
    ).scalars().first()


async def append_bundle(
    db: AsyncSession,
    *,
    user_id: str,
    ed25519_public_key: bytes,
    x25519_public_key: bytes,
    ml_kem_encapsulation_key: bytes,
    bundle_signature: bytes,
) -> KeyBundleRecord:
    """Add one bundle to a user's chain. The caller commits.

    Republishing an identical bundle is not appended. A client that re-runs enrolment, or
    reconnects and publishes again, would otherwise fill its own history with entries that
    record no change -- and a history where most rows mean "nothing happened" is one
    nobody reads closely enough to notice the row that does.
    """
    previous = await latest_record(db, user_id)
    if (
        previous is not None
        and previous.ed25519_public_key == ed25519_public_key
        and previous.x25519_public_key == x25519_public_key
        and previous.ml_kem_encapsulation_key == ml_kem_encapsulation_key
    ):
        return previous

    seq = (previous.seq if previous else 0) + 1
    record = KeyBundleRecord(
        user_id=user_id,
        seq=seq,
        ed25519_public_key=ed25519_public_key,
        x25519_public_key=x25519_public_key,
        ml_kem_encapsulation_key=ml_kem_encapsulation_key,
        bundle_signature=bundle_signature,
        prev_hash=previous.entry_hash if previous else None,
        entry_hash=entry_hash(
            prev_hash=previous.entry_hash if previous else None,
            user_id=user_id,
            seq=seq,
            ed25519_public_key=ed25519_public_key,
            x25519_public_key=x25519_public_key,
            ml_kem_encapsulation_key=ml_kem_encapsulation_key,
        ),
    )
    db.add(record)
    return record


def verify_chain(records: list[KeyBundleRecord]) -> tuple[bool, str | None]:
    """Recompute a user's chain end to end.

    Returns `(ok, reason)`. This runs server-side for the API's own answer, but the same
    computation is repeated in the browser against the returned history -- a log the
    relay grades itself against is not evidence of anything.
    """
    expected_prev: bytes | None = None
    for index, record in enumerate(records, start=1):
        if record.seq != index:
            return False, f"sequence jumps to {record.seq} at position {index}"
        if record.prev_hash != expected_prev:
            return False, f"entry {record.seq} does not follow the previous one"
        recomputed = entry_hash(
            prev_hash=record.prev_hash,
            user_id=record.user_id,
            seq=record.seq,
            ed25519_public_key=record.ed25519_public_key,
            x25519_public_key=record.x25519_public_key,
            ml_kem_encapsulation_key=record.ml_kem_encapsulation_key,
        )
        if recomputed != record.entry_hash:
            return False, f"entry {record.seq} has been altered"
        expected_prev = record.entry_hash
    return True, None
